### DISCUSSION TITLE
Is it possible to get relative CIP labels?

### DISCUSSION THREAD

**Discussion by j-wags** [rdkit/rdkit#4731] [Q&A] [answered]
Hi,

I was wondering if it's possible to get relative CIP stereo labels from RDKit. For example, I can get absolute stereo labels for 1-chloro,3-fluoro cyclohexane using the following:
```python
from rdkit import Chem
rdmol = Chem.MolFromSmiles("[C@@H]1(Cl)C[C@H](F)CCC1")  
Chem.AssignStereochemistry(rdmol)
for rda in rdmol.GetAtoms(): 
    if rda.HasProp("_CIPCode"):
        print(rda.GetProp("_CIPCode"))

```
> S
> R

![image](https://user-images.githubusercontent.com/16329175/143320633-bfa96fb6-a796-40bc-9aff-df323a388cb7.png)

But, if I shift the substituents to be para- to each other (1-chloro,4-fluoro cyclohexane) then absolute CIP stereochemistry doesn't work, and we'd need to fall back to relative CIP stereo (lower case `s` and `r`). However the code above doesn't work for that case.

```python
from rdkit import Chem
rdmol = Chem.MolFromSmiles("[C@@H]1(Cl)CC[C@H](F)CC1")  
Chem.AssignStereochemistry(rdmol)
for rda in rdmol.GetAtoms(): 
    if rda.HasProp("_CIPCode"):
        print(rda.GetProp("_CIPCode"))
```
> (no output)
![image](https://user-images.githubusercontent.com/16329175/143320784-7e7a9d87-8338-4317-90e6-d0b2d065bf2a.png)

Is there a code path to access relative CIP descriptors in this case? I'd have given up on it (since relative stereo looks [really complicated](http://ursula.chem.yale.edu/~chem220/chem220js/STUDYAIDS/isomers/RS14272/achiral.html#dioxa55)) and so it may not be implemented at all, but I do see hints of relative stereo functionality elsewhere in RDKit:
* This issue post on the same topic has "lower case (s)" in the molecule drawing in the third image: https://github.com/rdkit/rdkit/issues/4155#issuecomment-851923218
* The drawing code mentions relative cip: https://github.com/rdkit/rdkit/blob/b208da471f8edc88e07c77ed7d7868649ac75100/Code/GraphMol/MolDraw2D/MolDraw2DDetails.cpp#L50

(Also, some keywords that I'd searched before, in case it helps future people with the same question: relative stereo cip ring stereochemistry pseudosymmetry pseudoasymmetry achiral achirality)

**Comment by greglandrum** [comment, marked as answer]
Hi @j-wags, 
You need to use the new stereochemistry and CIP assignment code for this.
The incantation is `Chem.AssignCIPLabels()`:
```
In [7]: m = Chem.MolFromSmiles("[C@@H]1(Cl)CC[C@H](F)CC1")

In [8]: Chem.AssignCIPLabels(m)

In [9]: for rda in m.GetAtoms():
   ...:     if rda.HasProp("_CIPCode"):
   ...:         print(rda.GetProp("_CIPCode"))
   ...: 
s
s
```
There's an easier way:
```
In [5]: Chem.FindMolChiralCenters(m,useLegacyImplementation=False,force=True)
Out[5]: [(0, 's'), (4, 's')]
```

@ricrogz  and I did a pair of connected presentations about this at the 2020 UGM:
https://github.com/rdkit/UGM_2020/blob/master/Presentations/Landrum_Stereochem_Intro.pdf
https://github.com/rdkit/UGM_2020/blob/master/Presentations/RicardoRodriguezSchmidt_NewCIPLabelingCode.pdf

Since we tend to be fairly conservative (perhaps overly conservative) about trying to maintain behavior from one release to the next, this isn't yet the default.

**Reply by j-wags** [reply]
Wow! This works perfectly. Thanks so much for the quick response and additional context, @greglandrum! 

I only realized I missed the UGM this year after it passed - I'm really hoping to finally make it to the next one.
